README.md
<p align="center">
  <a href="https://github.com/homebridge/homebridge"><img src="https://raw.githubusercontent.com/homebridge/branding/master/logos/homebridge-color-round-stylized.png" height="140"></a>
</p>

<span align="center">

# homebridge-people-pro

[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins) [![npm](https://img.shields.io/npm/v/homebridge-people-pro.svg)](https://www.npmjs.com/package/homebridge-people-pro) [![npm](https://img.shields.io/npm/dt/homebridge-people-pro.svg)](https://www.npmjs.com/package/homebridge-people-pro) [![github](https://img.shields.io/github/last-commit/mfkrause/homebridge-people-pro)](https://www.github.com/mfkrause/homebridge-people-pro)

</span>

This is a plugin for [homebridge](https://github.com/homebridge/homebridge). It monitors who is at home, based on their smartphone being seen on the network recently.
If you use the Elgato Eve app you can also see the history of every person sensor (powered by [fakegato](https://github.com/simont77/fakegato-history]) - only works for sensors configured as "motion" sensors).

It can also optionally spin up a webserver and receive webhooks sent by location-aware mobile apps (such as [Locative](https://my.locative.io)), which can use iBeacons and geofencing to provide faster and more accurate location information.

# Installation

1.  Install homebridge (if not already installed) using: `npm install -g homebridge`
2.  Install this plugin using: `npm install -g homebridge-people-pro`
3.  Update your configuration file (see below).

# Configuration

See `config-sample.json` for an example configuration. This plugin can also be configured through a GUI like [homebridge-config-ui-x](https://github.com/oznu/homebridge-config-ui-x).

## Platform Configuration

| Parameter                  | Note                                                                                                                                                                                                |
| -------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `anyoneSensor`             | optional, default: false                                                                                                                                                                             |
| `anyoneSensorName`         | optional, default: "Anyone"                                                                                                                                                                         |
| `anyoneSensorType`         | optional, default: "motion", can be one of "motion", "occupancy".                                                                                                                                   |
| `nooneSensor`              | optional, default: false                                                                                                                                                                            |
| `nooneSensorName`          | optional, default: "No One"                                                                                                                                                                         |
| `nooneSensorType`          | optional, default: "motion", can be one of "motion", "occupancy".                                                                                                                                   |
| `webhookEnabled`           | optional, default: false, enable webhook functionality / webserver                                                                                                                                  |
| `webhookPort`              | optional, default: 51828                                                                                                                                                                            |
| `people`                   | array of objects of the sensors / people to set-up, see below for configuration of every sensor                                                                                                     |

## Sensors / People Configuration

| Parameter              | Note                                                                                                                                                                                                |
| ---------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `target`               | may be either a hostname, IP address or MAC address                                                                                                                                                 |
| `customDns`            | if `target` is a hostname, you can optionally set an array of custom DNS servers here to use for the hostname lookup.                                                                               |
| `name`                 | a human-readable name for your sensor                                                                                                                                                               |
| `type`                 | optional, default: "motion", can be one of "motion", "occupancy". **WARNING**: Choosing something else than "motion" will disable fakegato / Elgato Eve history functionality.                      |
| `threshold`            | optional, in minutes, default: 15                                                                                                                                                                   |
| `pingInterval`         | optional, in milliseconds, default: 10000, if set to -1 the ping/arp mechanism will not be used                                                                                                     |
| `pingUseArp`           | optional, default: false, use ARP lookup tables instead of ICMP ping                                                                                                                                |
| `ignoreWebhookReEnter` | optional, in seconds, default: 0, if set to 0 every webhook re-enter/exit will trigger state change; otherwise the state will only change if no re-enter/exit occurs in specified number of seconds |
| `excludeFromWebhook`   | optional, default: false, if set to true, this sensor won't be able to be managed through webhooks / will ignore webhook requests                                                                   |

# How it works

-   When started, homebridge-people-pro will continually ping (or ARP lookup if configured) the IP address associated with each person defined in config.json if `pingInterval` is not set to `-1`. If a hostname is given as a target, you can also define custom DNS servers to use for the look-up. If a MAC address is given as a target, the plugin will try to look-up its IP address through ARP before pinging.
-   With an iBeacon or geofencing smartphone app, you can configure a HTTP push to trigger when you enter and exit your 'home' region. This data will be combined with the ping functionality if used to give this plugin more precise presence data.
-   When a ping is successful, the current timestamp is logged to a database.
-   When a HomeKit enabled app looks up the state of a person, the last seen time for that persons device is compared to the current time minus `threshold` minutes, and if it is greater it is assumed that the person is active.

## 'Anyone' and 'No One' sensors

Some HomeKit automations need to happen when "anyone" is home or when "no one" is around. homebridge-people-pro can automatically create additional "Anyone" and "No One" sensors to make these automations very easy.

For example, you might want to run your "Arrive Home" scene when *Anyone* gets home. Or run "Leave Home" when *No One* is home.

These sensors can be enabled by adding `"anyoneSensor" : true` and/or `"nooneSensor" : true` to your homebridge `config.json` file.

## Accuracy

This plugin requires that the devices being monitored are connected to the network. iPhones (and other devices) may deliberately disconnect from the network once the screen is turned off to save power, meaning just because the device isn't connected, it doesn't mean that the devices owner isn't at home. Fortunately, most of these devices periodically reconnect to the network to check for updates, emails, etc. This plugin works by keeping track of the last time a device was seen, and comparing that to a threshold value (in minutes).

From a *very* limited amount of testing, it was found that a threshold of 15 minutes seems to work well for most phones. The threshold can be configured individually for each sensor.

Additionally, if you're using a location-aware mobile app to range for iBeacons and geofences, this plugin can receive a HTTP push from the app to immediately see you as present or not present when you physically enter or exit your desired region. This is particularly useful for "Arrive Home" and "Depart Home" HomeKit automations which ideally happen faster than every 15 minutes.

## Pairing with a location-aware mobile app

Apps like [Locative](https://my.locative.io) range for iBeacons and geofences by using core location APIs available on your smartphone. With Bluetooth and location services turned on, these apps can provide an instantaneous update when you enter and exit a desired region.

To use this plugin with one of these apps, configure your region and set the HTTP push to `http://[ipaddress]:[port]/?sensor=[name]&state=true` for arrival, and `http://[ipaddress]:[port]/?sensor=[name]&state=false` for departure, where `[ipaddress]` is the IP address of your Homebridge, `[port]` the configured port for the webhooks (defaults to 51828) and `[name]` the name of the person the device belongs to as specified in your config under `people`. *Note:* You may need to enable port forwarding on your router to accomplish this.

# Notes

## Running on a raspberry pi as non 'pi' user

On some versions of raspbian, users are not able to use the ping program by default. If none of your devices show online try running `sudo chmod u+s /bin/ping`. Thanks to oberstmueller for the tip.

## Running in a docker-environment

On some docker-environments (alpine-based for example) it is possible that the ping does not. Please try to install *iptools* in this case via `apk add iputils --no-cache`

## Thanks

Thanks to everyone who's helped contribute code, feedback and support. In particular:

-   [PeteLawrence](https://github.com/PeteLawrence/homebridge-people) - for the original plugin
-   [simont77](https://github.com/simont77/fakegato-history) - for the fakegato-plugin
-   [wr](https://github.com/wr) - for adding in webhook support
-   [benzman81](https://github.com/benzman81) - for porting the plugin over to be a Platform and improving how ping and webhooks work together, and numerous other fixes
-   [skrollme](https://github.com/skrollme) - for the plugin this one is forked from


CHANGELOG.md
# Changelog

## 0.11.5
- [Security] Bumped `get-ip-range` to `^4.0.0` due to dependabot security alert (DoS)
## 0.11.4
- [Bug] Fixes a bug where occupancy sensors wouldn't update immediately after a ping/arp status change.

## 0.11.1
- [Improvement] :warning: anyoneSensor now defaults to false.
- [Bug] Fixes an unhandled exception on plug-in start when an invalid or incomplete config was given.
- [Bug] Fixes typo on the excludeFromWebhook config option.
- [Bug] Fixes several default config values.

## 0.11.0
- [Feature] You can now use a MAC address as a target. The corresponding IP address will be looked up through ARP before each sensor update (#9).

## 0.10.0
- [Feature] Adds option to add custom DNS server(s) to use for hostname look-ups prior to pinging (#8)

## 0.9.3
- [Bug] Fixed last activation callback handler when the sensor has not been activated in the past - fixes a warning message about the plugin slowing down Homebridge in Homebridge >= v1.3.1.

## 0.9.2
- [Improvement] "Anyone" and "No One" sensors can now also be configured as occupancy sensors.
- [Bug/Improvement] Several stability optimizations and bugfixes.

## 0.9.1
- [Feature] Sensors can now optionally be occupancy sensors instead of motion sensors (however this will disable fakegato support)
- [Improvement] Completely removed cacheDirectory config, defaulting to the Homebridge storage path
- [Bug] Fixed plugin start-up on macOS

## 0.9.0
- [Feature] "Anyone" and "No One" sensors can now be renamed
- [Feature] Added config schema for GUI configuration
- [Improvement] Webhooks are now an optional configuration so that the webserver is not exposed unnecessarily
- [Improvement] Specific targets / sensors can now be excluded from webhook functionality
- [Improvement] Renamed ignoreReEnterExitSeconds option to ignoreWebhookReEnter to make functionality more clear
- [Code] Added documentation in code

## 0.8.0
- [Feature] Added "pingUseArp" configuration
- [Code] Added ESLint for better code style
- [Code] Refactored complete codebase


src/platform.js
const storage = require('node-persist');
const http = require('http');
const url = require('url');

const PeopleProAccessory = require('./accessory');
const PeopleProAllAccessory = require('./all_accessory');

let homebridge;

class PeopleProPlatform {
  constructor(log, config) {
    this.log = log;
    this.config = config;
    this.threshold = config.threshold || 15;
    this.anyoneSensor = ((typeof (config.anyoneSensor) !== 'undefined' && config.anyoneSensor !== null) ? config.anyoneSensor : false);
    this.nooneSensor = ((typeof (config.nooneSensor) !== 'undefined' && config.nooneSensor !== null) ? config.nooneSensor : false);
    this.anyoneSensorName = config.anyoneSensorName || 'Anyone';
    this.nooneSensorName = config.nooneSensorName || 'No One';
    this.webhookPort = config.webhookPort || 51828;
    this.webhookEnabled = ((typeof (config.webhookEnabled) !== 'undefined' && config.webhookEnabled !== null) ? config.webhookEnabled : false);
    this.pingInterval = config.pingInterval || 10000;
    this.ignoreWebhookReEnter = config.ignoreWebhookReEnter || 0;
    this.people = config.people || [];
    this.storage = storage;
    this.storage.initSync({ dir: `${homebridge.user.storagePath()}/plugin-persist/homebridge-people-pro` });
    this.webhookQueue = [];
  }

  accessories(callback) {
    this.accessories = [];
    this.peopleProAccessories = [];

    // Get all people / targets and add them to accessories
    for (let i = 0; i < this.people.length; i += 1) {
      const peopleProAccessory = new PeopleProAccessory(this.log, this.people[i], this);
      this.accessories.push(peopleProAccessory);
      this.peopleProAccessories.push(peopleProAccessory);
    }

    // Add "anyone" and "no one" sensors / accessories
    if (this.anyoneSensor) {
      this.peopleAnyOneAccessory = new PeopleProAllAccessory(
        this.log, this.config, this.anyoneSensorName, this, 'anyone',
      );
      this.accessories.push(this.peopleAnyOneAccessory);
    }
    if (this.nooneSensor) {
      this.peopleNoOneAccessory = new PeopleProAllAccessory(
        this.log, this.config, this.nooneSensorName, this, 'noone',
      );
      this.accessories.push(this.peopleNoOneAccessory);
    }
    callback(this.accessories);

    // Start webhook server if enabled
    if (this.webhookEnabled) {
      this.startServer();
    }
  }

  /**
   *  Spins up a webserver for the webhook functionality.
   *
   *  HTTP webserver code influenced by benzman81's great homebridge-http-webhooks plugin:
   *  https://github.com/benzman81/homebridge-http-webhooks
   */
  startServer() {
    http.createServer(((request, response) => {
      const theUrl = request.url;
      const theUrlParts = url.parse(theUrl, true);
      const theUrlParams = theUrlParts.query;
      let body = [];
      request.on('error', ((err) => {
        this.log('Webhook error: %s.', err);
      })).on('data', (chunk) => {
        body.push(chunk);
      }).on('end', (() => {
        body = Buffer.concat(body).toString();

        response.on('error', (err) => {
          this.log('Webhook error: %s.', err);
        });

        response.statusCode = 200;
        response.setHeader('Content-Type', 'application/json');

        if (!theUrlParams.sensor || !theUrlParams.state) {
          // Received invalid request
          response.statusCode = 404;
          response.setHeader('Content-Type', 'text/plain');
          const errorText = 'Webhook error: No sensor or state specified in request.';
          this.log(errorText);
          response.write(errorText);
          response.end();
        } else {
          const sensor = theUrlParams.sensor.toLowerCase();
          const newState = (theUrlParams.state === 'true');
          this.log(`Received webhook for ${sensor} -> ${newState}`);
          const responseBody = {
            success: true,
          };

          // Loop through sensors to find which one to update
          // Will always return 200, even if sensor can't be found due to security reasons
          for (let i = 0; i < this.peopleProAccessories.length; i += 1) {
            const peopleProAccessory = this.peopleProAccessories[i];
            const { target } = peopleProAccessory;
            if (peopleProAccessory.name.toLowerCase() === sensor) {
              // Check if this sensor is excluded from webhook functionality; if so, ignore request
              if (peopleProAccessory.excludedFromWebhook !== true) {
                // Update webhook queue
                this.clearWebhookQueueForTarget(target);
                this.webhookQueue.push({
                  target,
                  newState,
                  timeoutvar: setTimeout((() => {
                    this.runWebhookFromQueueForTarget(target);
                  }), peopleProAccessory.ignoreWebhookReEnter * 1000),
                });
              }
              break;
            }
          }
          response.write(JSON.stringify(responseBody));
          response.end();
        }
      }));
    })).listen(this.webhookPort);
    this.log("Webhook: Started webserver on port '%s'.", this.webhookPort);
  }

  /**
   * Clears the current webhook queue / intervals for the given target
   * @param {string} target The target to clear the webhook queue for
   */
  clearWebhookQueueForTarget(target) {
    for (let i = 0; i < this.webhookQueue.length; i += 1) {
      const webhookQueueEntry = this.webhookQueue[i];
      if (webhookQueueEntry.target === target) {
        clearTimeout(webhookQueueEntry.timeoutvar);
        this.webhookQueue.splice(i, 1);
        break;
      }
    }
  }

  /**
   * Executes the webhook in the queue for the given target; called through an interval at
   * ignoreWebhookReEnter * 1000 from the webserver
   * @param {string} target The target to run the webhook for
   */
  runWebhookFromQueueForTarget(target) {
    for (let i = 0; i < this.webhookQueue.length; i += 1) {
      const webhookQueueEntry = this.webhookQueue[i];
      if (webhookQueueEntry.target === target) {
        this.log(`Running webhook for ${target} -> ${webhookQueueEntry.newState}`);
        this.webhookQueue.splice(i, 1);
        // Update sensor
        this.storage.setItemSync(`lastWebhook_${target}`, Date.now());
        this.getPeopleProAccessoryForTarget(target).setNewState(webhookQueueEntry.newState);
        break;
      }
    }
  }

  /**
   * Get a PeopleProAccessory based on a given target
   * @param {string} target IP address or hostname of the wanted PeopleProAccessory
   * @returns {object} The wanted PeopleProAccessory (or null if it couldn't be found)
   */
  getPeopleProAccessoryForTarget(target) {
    for (let i = 0; i < this.peopleProAccessories.length; i += 1) {
      const peopleProAccessory = this.peopleProAccessories[i];
      if (peopleProAccessory.target === target) {
        return peopleProAccessory;
      }
    }
    return null;
  }
}

/**
 * Set homebridge reference for platform, called from /index.js
 * @param {object} homebridgeRef The homebridge reference to use in the platform
 */
PeopleProPlatform.setHomebridge = (homebridgeRef) => {
  homebridge = homebridgeRef;
};

module.exports = PeopleProPlatform;
